### ISSUE TITLE
Service can't reach it's own vip/dns name

### ISSUE THREAD

**Issue by tonistiigi** [docker/docker#24780]
Single node:

```
docker network create mynet -d overlay
docker service create --name foo --network mynet --replicas 3 busybox nc -ll -p 1234 -e hostname
docker service create --name bar --network mynet --replicas 1 busybox top
```

Accessing service `foo` from other service works all the time:
```
docker exec $(docker ps | grep bar.1 | awk '{print $1}') nc foo:1234
```

Accessing from `foo` itself only works when vip resolves to current container, 1/3 cases for this example:
```
docker exec $(docker ps | grep foo.1 | awk '{print $1}') nc foo:1234
```

@aboch @mavenugo

**Comment by mavenugo** [comment]
@tonistiigi can you confirm if you are running latest master which contains https://github.com/docker/docker/pull/24573 ?

**Comment by tonistiigi** [comment]
@mavenugo Just checked with e6a97db2c932a2b80180c10f2896d7594108a406 , same behavior

**Comment by mavenugo** [comment]
thanks @tonistiigi I could reproduce the issue as well. I will think of a better fix.

**Comment by kalahari** [comment]
Assigning the service name to the loopback interface via `/etc/hosts` seems like a pretty simple fix/work-around. Are there any down sides to that approach?

**Comment by sirlatrom** [comment]
@kalahari With some software which doesn't respect/care about `/etc/hosts`, this might at least create confusing situations where other services can be reached over DNS while 'peer' service instances will not be resolvable (due to not reading `/etc/hosts`), e.g. the way the `oraclelinux` works (or at least it used to be this way).

**Comment by sanimej** [comment]
@tonistiigi  Fix for this is in master and will be available in 1.12.1.
